src/ast/query.rs

```diff
@@ -1865,27 +1873,57 @@ impl fmt::Display for SelectInto {
     }
 }
 
+#[derive(Debug, Copy, Clone, PartialEq, PartialOrd, Eq, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "visitor", derive(Visit, VisitMut))]
+pub enum WithModifier {
+    Rollup,
+    Cube,
+    Totals,
+}
+
+impl fmt::Display for WithModifier {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            WithModifier::Rollup => write!(f, "WITH ROLLUP"),
+            WithModifier::Cube => write!(f, "WITH CUBE"),
+            WithModifier::Totals => write!(f, "WITH TOTALS"),
+        }
+    }
+}
+
 #[derive(Debug, Clone, PartialEq, PartialOrd, Eq, Ord, Hash)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "visitor", derive(Visit, VisitMut))]
 pub enum GroupByExpr {
-    /// ALL syntax of [Snowflake], and [DuckDB]
+    /// ALL syntax of [Snowflake], [DuckDB] and [ClickHouse].
     ///
     /// [Snowflake]: <https://docs.snowflake.com/en/sql-reference/constructs/group-by#label-group-by-all-columns>
     /// [DuckDB]:  <https://duckdb.org/docs/sql/query_syntax/groupby.html>
-    All,
+    /// [ClickHouse]: <https://clickhouse.com/docs/en/sql-reference/statements/select/group-by#group-by-all>
+    ///
+    /// ClickHouse also supports WITH modifiers after GROUP BY ALL and expressions.
+    ///
+    /// [ClickHouse]: <https://clickhouse.com/docs/en/sql-reference/statements/select/group-by#rollup-modifier>
+    All(Vec<WithModifier>),
 
     /// Expressions
-    Expressions(Vec<Expr>),
+    Expressions(Vec<Expr>, Vec<WithModifier>),
 }
 
 impl fmt::Display for GroupByExpr {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match self {
-            GroupByExpr::All => write!(f, "GROUP BY ALL"),
-            GroupByExpr::Expressions(col_names) => {
+            GroupByExpr::All(modifiers) => {
+                write!(f, "GROUP BY ALL")?;
+                write!(f, " {}", display_separated(modifiers, " "))?;
+                Ok(())
+            }
+            GroupByExpr::Expressions(col_names, modifiers) => {
                 let col_names = display_comma_separated(col_names);
-                write!(f, "GROUP BY ({col_names})")
+                write!(f, "GROUP BY ({col_names})")?;
+                write!(f, " {}", display_separated(modifiers, " "))?;
```

Comment:
In both clauses do we need to check that modifiers is not empty before displaying it? (Looks otherwise like we may get a trailing space in the output if there are no modifiers)